pkg/reconciler/namespace/namespace.go

```diff
@@ -187,111 +185,32 @@ func (c *Controller) updateScheduledCondition(ctx context.Context, ns *corev1.Na
-		if _, err = c.kubeClient.Cluster(ns.ClusterName).CoreV1().Namespaces().
-			Patch(ctx, ns.Name, patchType, patchBytes, metav1.PatchOptions{}); err != nil {
-			return err
-		}
-		klog.Infof("Patched cluster assignment for namespace %q|%q: %q -> %q", ns.ClusterName, ns.Name, oldClusterName, newClusterName)
-	}
-
-	return nil
-}
-
-// isValidCluster checks whether the given cluster name exists and is valid for
-// the purposes of any namespace already scheduled to it (i.e., if it reports
-// as Ready, and any evictAfter value, if specified, has not yet passed).
-//
-// It doesn't take into account Unschedulable, and should only be used when
-// determining if a cluster that a namespace has already been assigned to
-// should keep having that namespace.
-func (c *Controller) isValidCluster(ctx context.Context, lclusterName, clusterName string) (bool, error) {
-	cluster, err := c.clusterLister.Get(clusters.ToClusterAwareKey(lclusterName, clusterName))
-	if apierrors.IsNotFound(err) {
-		klog.Infof("Cluster %q|%q does not exist", lclusterName, clusterName)
-		return false, nil
-	}
-	if err != nil {
-		return false, err
-	}
-	if ready := conditions.IsTrue(cluster, clusterv1alpha1.ClusterReadyCondition); !ready {
-		klog.Infof("Cluster %q|%q is not reporting ready", lclusterName, clusterName)
-		return false, nil
-	}
-	if evictAfter := cluster.Spec.EvictAfter; evictAfter != nil && evictAfter.Time.Before(time.Now()) {
-		klog.Infof("Cluster %q|%q is evicted", lclusterName, clusterName)
-		return false, nil
-	}
-	return true, nil
-}
-
 // ensureScheduled attempts to ensure the namespace is assigned to a viable cluster. This
 // will succeed without error if a cluster is assigned or if there are no viable clusters
 // to assign to. The condition of not being scheduled to a cluster will be reflected in
 // the namespace's status rather than by returning an error.
-func (c *Controller) ensureScheduled(ctx context.Context, lclusterName string, ns *corev1.Namespace) error {
-	clusterName := ns.Labels[clusterLabel]
+func (c *Controller) ensureScheduled(ctx context.Context, ns *corev1.Namespace) error {
+	oldPClusterName := ns.Labels[clusterLabel]
 
-	noClusterAssigned := clusterName == ""
-	if noClusterAssigned {
-		if err := c.assignCluster(ctx, ns); err != nil {
-			return err
-		}
-	} else {
-		isValidCluster, err := c.isValidCluster(ctx, lclusterName, clusterName)
-		if err != nil {
-			return err
-		}
-		if !isValidCluster {
-			// The currently assigned cluster no longer exists or is not ready.
-			if err := c.assignCluster(ctx, ns); err != nil {
-				return err
-			}
-		}
+	scheduler := namespaceScheduler{
+		getCluster:   c.clusterLister.Get,
+		listClusters: c.clusterLister.List,
+	}
+	newPClusterName, err := scheduler.AssignCluster(ns)
+	if err != nil {
+		return err
 	}
 
-	return nil
+	if oldPClusterName == newPClusterName {
+		return nil
+	}
+
+	klog.Infof("Patching to update cluster assignment for namespace %q|%q: %q -> %q",
```

Comment:
Discussion in Slack sounded like it was heading toward `%s` over `%q`

Reply:
This isn't net-new code - it's a refactor - and I'd much prefer to defer rewriting log messages to a separate non-functional PR.

Reply:
https://github.com/kcp-dev/kcp/pull/600 does the rename.